crates/protocol/src/io.rs

```diff
@@ -0,0 +1,301 @@
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        Ok(buffer.read_u8()? != 0)
+    }
+}
+
+impl Encoder for bool {
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()> {
+        buffer.write_u8(if *self { 1 } else { 0 })?;
+        Ok(())
+    }
+}
+
+const STRING_MAX_LENGTH: usize = 32767;
+
+impl Decoder for String {
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        let len = VarInt::decode(buffer)
+            .context("failed to decode string length")?
+            .0 as usize;
+
+        if len > STRING_MAX_LENGTH {
+            bail!(
+                "string length {} exceeds maximum allowed length of {}",
+                len,
+                STRING_MAX_LENGTH
+            );
+        }
+
+        let mut bytes = vec![0u8; len];
+        buffer
+            .read_exact(&mut bytes)
+            .map_err(|_| Error::UnexpectedEof("String"))?;
+
+        let str = std::str::from_utf8(&bytes).context("failed to decode string")?;
+
+        Ok(str.to_owned())
+    }
+}
+
+impl Encoder for String {
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()> {
+        let bytes = self.as_bytes();
+        if bytes.len() > STRING_MAX_LENGTH {
+            bail!(
+                "string length {} exceeds maximum allowed length of {}",
+                bytes.len(),
+                STRING_MAX_LENGTH
+            );
+        }
+
+        VarInt(bytes.len() as i32).encode(buffer)?;
+        buffer.extend_from_slice(bytes);
+        Ok(())
+    }
+}
+
+impl Decoder for Uuid {
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        let mut bytes = [0u8; 16];
+        buffer
+            .read_exact(&mut bytes)
+            .map_err(|_| Error::UnexpectedEof("Uuid"))?;
+
+        Ok(Uuid::from_bytes(bytes))
+    }
+}
+
+impl Encoder for Uuid {
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()> {
+        buffer.extend_from_slice(self.as_bytes());
+        Ok(())
+    }
+}
+
+impl<T> Decoder for Option<T>
+where
+    T: Decoder,
+{
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self> {
+        let present = bool::decode(buffer)?;
+        if present {
+            Ok(Some(T::decode(buffer)?))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<T> Encoder for Option<T>
+where
+    T: Encoder,
+{
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()> {
+        match self {
+            Some(value) => {
+                true.encode(buffer)?;
+                value.encode(buffer)?;
+            }
+            None => {
+                false.encode(buffer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("unexpected end of input: failed to read value of type `{0}`")]
+    UnexpectedEof(&'static str),
+}
+
+const MAX_LENGTH: usize = 1024 * 1024; // 2^20 elements
+
+pub struct LengthPrefixedVec<'a, P, T>(pub Cow<'a, [T]>, PhantomData<P>)
```

Comment:
At least implement `Debug`.

And probably some other convenience traits.

Reply:
Outdated